// quicklyzer/dist/services/project.js
import { join } from "node:path";
import { fileExists, readJsonFile } from "../utils/system.js";
import { getProjectStatistics } from "./statistics.js";
function detectPackageManager(projectPath) {
    if (fileExists(join(projectPath, "package-lock.json"))) {
        return "npm";
    }
    if (fileExists(join(projectPath, "pnpm-lock.yaml"))) {
        return "pnpm";
    }
    if (fileExists(join(projectPath, "yarn.lock"))) {
        return "yarn";
    }
    if (fileExists(join(projectPath, "bun.lockb"))) {
        return "bun";
    }
    return "Unknown";
}
function detectLanguage(projectPath) {
    if (fileExists(join(projectPath, "tsconfig.json"))) {
        return "TypeScript";
    }
    if (fileExists(join(projectPath, "jsconfig.json"))) {
        return "JavaScript";
    }
    return "Unknown";
}
function detectFramework(pkg) {
    const deps = {
        ...(pkg.dependencies ?? {}),
        ...(pkg.devDependencies ?? {})
    };
    if ("next" in deps)
        return "Next.js";
    if ("react" in deps)
        return "React";
    if ("vue" in deps)
        return "Vue";
    if ("@angular/core" in deps)
        return "Angular";
    if ("express" in deps)
        return "Express";
    if ("@nestjs/core" in deps)
        return "NestJS";
    if ("svelte" in deps)
        return "Svelte";
    return "Unknown";
}
function detectFrameworkVersion(pkg) {
    const deps = {
        ...(pkg.dependencies ?? {}),
        ...(pkg.devDependencies ?? {})
    };
    const frameworks = [
        "next",
        "react",
        "vue",
        "@angular/core",
        "express",
        "@nestjs/core",
        "svelte",
    ];
    for (const framework of frameworks) {
        if (framework in deps) {
            return deps[framework];
        }
    }
    return "Unknown";
}
function detectBuildTool(pkg) {
    const deps = {
        ...(pkg.dependencies ?? {}),
        ...(pkg.devDependencies ?? {})
    };
    if ("vite" in deps)
        return "Vite";
    if ("webpack" in deps)
        return "Webpack";
    if ("rollup" in deps)
        return "Rollup";
    if ("parcel" in deps)
        return "Parcel";
    if ("@rspack/core" in deps)
        return "Rspack";
    if ("turbopack" in deps)
        return "Turbopack";
    return "Unknown";
}
function detectBuildToolVersion(pkg) {
    const deps = {
        ...(pkg.dependencies ?? {}),
        ...(pkg.devDependencies ?? {})
    };
    const buildTools = [
        "vite",
        "webpack",
        "rollup",
        "parcel",
        "@rspack/core",
        "turbopack",
        "esbuild",
        "swc",
        "snowpack",
        "brunch",
        "gulp",
        "grunt",
        "broccoli",
        "fuse-box"
    ];
    for (const tool of buildTools) {
        if (tool in deps) {
            return deps[tool];
        }
    }
    return "Unknown";
}
function getDependencyStatistics(pkg) {
    const dependencies = pkg.dependencies ?? {};
    const devDependencies = pkg.devDependencies ?? {};
    const dependencyCount = Object.keys(dependencies).length;
    const devDependencyCount = Object.keys(devDependencies).length;
    return {
        dependencyCount,
        devDependencyCount,
        totalDependencyCount: dependencyCount + devDependencyCount
    };
}
function getScripts(pkg) {
    return Object.keys(pkg.scripts ?? {});
}
function detectNodeVersion(pkg) {
    return pkg.engines?.node ?? "Not specified";
}
function detectDocker(projectPath) {
    const dockerFiles = [
        "Dockerfile",
        "docker-compose.yml",
        "docker-compose.yaml",
        "compose.yml",
        "compose.yaml",
        ".dockerignore"
    ];
    return dockerFiles.some(file => fileExists(join(projectPath, file)));
}
function detectCI(projectPath) {
    if (fileExists(join(projectPath, ".github", "workflows"))) {
        return "GitHub Actions";
    }
    if (fileExists(join(projectPath, ".gitlab-ci.yml"))) {
        return "GitLab CI";
    }
    if (fileExists(join(projectPath, "azure-pipelines.yml"))) {
        return "Azure Pipelines";
    }
    if (fileExists(join(projectPath, "circle.yml"))) {
        return "CircleCI";
    }
    return "None";
}
function detectESLint(projectPath, pkg) {
    const eslintFiles = [
        "eslint.config.js",
        "eslint.config.mjs",
        "eslint.config.cjs",
        ".eslintrc",
        ".eslintrc.js",
        ".eslintrc.cjs",
        ".eslintrc.json",
        ".eslintrc.yml",
        ".eslintrc.yaml"
    ];
    const hasConfig = eslintFiles.some(file => fileExists(join(projectPath, file)));
    const deps = {
        ...(pkg.dependencies ?? {}),
        ...(pkg.devDependencies ?? {})
    };
    return hasConfig || "eslint" in deps;
}
function detectPrettier(projectPath, pkg) {
    const prettierFiles = [
        ".prettierrc",
        ".prettierrc.json",
        ".prettierrc.js",
        ".prettierrc.cjs",
        ".prettierrc.yml",
        ".prettierrc.yaml",
        "prettier.config.js",
        "prettier.config.cjs"
    ];
    const hasConfig = prettierFiles.some(file => fileExists(join(projectPath, file)));
    const deps = {
        ...(pkg.dependencies ?? {}),
        ...(pkg.devDependencies ?? {})
    };
    return hasConfig || "prettier" in deps;
}
function detectMonorepo(projectPath, pkg) {
    const workspaceFiles = [
        "pnpm-workspace.yaml",
        "pnpm-workspace.yml",
        "turbo.json",
        "nx.json"
    ];
    const hasWorkspaceFile = workspaceFiles.some(file => fileExists(join(projectPath, file)));
    const hasWorkspaces = Array.isArray(pkg.workspaces) &&
        pkg.workspaces.length > 0;
    return hasWorkspaceFile || hasWorkspaces;
}
function detectGit(projectPath) {
    return fileExists(join(projectPath, ".git"));
}
function detectReadme(projectPath) {
    return (fileExists(join(projectPath, "README.md")) ||
        fileExists(join(projectPath, "README.MD")) ||
        fileExists(join(projectPath, "readme.md")) ||
        fileExists(join(projectPath, "readme.MD")) ||
        fileExists(join(projectPath, "Readme.md")) ||
        fileExists(join(projectPath, "Readme.MD")));
}
function detectLicense(projectPath) {
    return (fileExists(join(projectPath, "LICENSE")) ||
        fileExists(join(projectPath, "LICENSE.md")) ||
        fileExists(join(projectPath, "LICENSE.txt")) ||
        fileExists(join(projectPath, "license")) ||
        fileExists(join(projectPath, "license.md")) ||
        fileExists(join(projectPath, "license.txt")) ||
        fileExists(join(projectPath, "License")) ||
        fileExists(join(projectPath, "License.md")) ||
        fileExists(join(projectPath, "License.txt")));
}
export function analyzeProject() {
    const projectPath = process.cwd();
    const packageJsonPath = join(projectPath, "package.json");
    if (!fileExists(packageJsonPath)) {
        throw new Error("No package.json found in the current directory.");
    }
    const pkg = readJsonFile(packageJsonPath);
    const dependencyStats = getDependencyStatistics(pkg);
    const scripts = getScripts(pkg);
    const nodeVersion = detectNodeVersion(pkg);
    const docker = detectDocker(projectPath);
    const ci = detectCI(projectPath);
    const eslint = detectESLint(projectPath, pkg);
    const prettier = detectPrettier(projectPath, pkg);
    const monorepo = detectMonorepo(projectPath, pkg);
    const statistics = getProjectStatistics(projectPath);
    return {
        name: pkg.name ?? "Unknown",
        version: pkg.version ?? "Unknown",
        packageManager: detectPackageManager(projectPath),
        language: detectLanguage(projectPath),
        framework: detectFramework(pkg),
        frameworkVersion: detectFrameworkVersion(pkg),
        buildTool: detectBuildTool(pkg),
        buildToolVersion: detectBuildToolVersion(pkg),
        dependencyCount: dependencyStats.dependencyCount,
        devDependencyCount: dependencyStats.devDependencyCount,
        totalDependencyCount: dependencyStats.totalDependencyCount,
        totalFiles: statistics.totalFiles,
        sourceFiles: statistics.sourceFiles,
        directories: statistics.directories,
        largestDirectory: statistics.largestDirectory,
        linesOfCode: statistics.linesOfCode,
        largestFile: statistics.largestFile,
        emptyDirectories: statistics.emptyDirectories,
        hiddenFiles: statistics.hiddenFiles,
        projectSize: statistics.projectSize,
        scripts: scripts,
        nodeVersion: nodeVersion,
        docker: docker,
        ci: ci,
        eslint: eslint,
        prettier: prettier,
        monorepo: monorepo,
        git: detectGit(projectPath),
        readme: detectReadme(projectPath),
        license: detectLicense(projectPath)
    };
}
